syft/execution/role.py

```diff
@@ -312,6 +313,100 @@ def copy(self):
             id=sy.ID_PROVIDER.pop(),
         )
 
+    def _is_inplace_operation(self, cmd: str, target: PlaceholderId):
+        """
+        Helper method that returns True if cmd is inplace operation
+        """
+        if target is None:
+            framework_name, command = cmd.split(".", 2)
+            framework = getattr(syft, framework_name, syft.framework)
+            return framework.is_inplace_method(command)
+
+        elif isinstance(target, PlaceholderId):
+            ph = self.placeholders.get(target.value, None)
+            if ph is not None:
+                framework_name = ph.child.__module__
+                framework = getattr(syft, framework_name, syft.framework)
+                return framework.is_inplace_method(cmd)
+
+        return False
+
+    def _prune_actions(self):
+        """
+        Removes unnecessary actions and placeholders.
+        """
+
+        # Track operations on output placeholders from bottom to top
+        output_connected_placeholder_ids = set(self.output_placeholder_ids)
+        connected_actions_idx = set()
+        for action_rev_idx, action in enumerate(reversed(self.actions)):
+            return_ids = (
+                {ph.value for ph in action.return_ids if isinstance(ph, PlaceholderId)}
+                if action.return_ids is not None
+                else set()
+            )
+            target_id = action.target.value if action.target is not None else None
+            # Operation resulted in placeholder connected to output
+            returns_connected_ph = (
+                len(return_ids.intersection(output_connected_placeholder_ids)) > 0
+            )
+            # Operation has side-effect on placeholder connected to output
+            affects_connected_ph = (
+                target_id in output_connected_placeholder_ids
+                and self._is_inplace_operation(action.name, action.target)
+            )
+
+            if returns_connected_ph or affects_connected_ph:
+                connected_actions_idx.add(len(self.actions) - 1 - action_rev_idx)
```

Comment:
Could you add a line to explain why we don't use action_rev_idx (as we use action_idx below) ? (I guess, because we did a reverse enumerate)